src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,551 @@
+			tdfb_cdec_s16(&x, cd->direction.d, cd->direction.d_size);
+			r = 0;
+			for (i = 0; i < frames; i++) {
+				if (maxval < *y)
+					maxval = *y;
+
+				r += (int32_t)*x * *y;
+				y += ch_count;
+				tdfb_cinc_s16(&y, cd->direction.d_end, cd->direction.d_size);
+				x += ch_count;
+				tdfb_cinc_s16(&x, cd->direction.d_end, cd->direction.d_size);
+			}
+			/* Scale for max. 20 ms 48 kHz frame */
+			cd->direction.r[k + max_lag] = sat_int32(((r >> 8) + 1) >> 1);
+		}
+		r_max_idx = find_max_value_index(&cd->direction.r[0], 2 * max_lag + 1);
+		cd->direction.timediff[c - 1] = (int32_t)(r_max_idx - max_lag) *
+			cd->direction.unit_delay;
+	}
+	cd->direction.rp += frames * ch_count;
+	tdfb_cinc_s16(&cd->direction.rp, cd->direction.d_end, cd->direction.d_size);
+}
+
+static int16_t distance_from_source(struct tdfb_comp_data *cd, int mic_n,
+				    int16_t x, int16_t y, int16_t z)
+{
+	int32_t d2;
+	int16_t dx;
+	int16_t dy;
+	int16_t dz;
+	int16_t d;
+
+	dx = x - cd->mic_locations[mic_n].x;
+	dy = y - cd->mic_locations[mic_n].y;
+	dz = z - cd->mic_locations[mic_n].z;
+
+	/* d2 is Q8.24 meters */
+	d2 = dx * dx + dy * dy + dz * dz;
+
+	/* Squared distance is Q8.24, return Q4.12 meters */
+	d = tdfb_mic_distance_sqrt(d2);
+	return d;
+}
+
+static inline void theoretical_time_differences(struct tdfb_comp_data *cd, int16_t az)
+{
+	int16_t d[PLATFORM_MAX_CHANNELS];
+	int16_t src_x;
+	int16_t src_y;
+	int16_t sin_az;
+	int16_t cos_az;
+	int32_t delta_d;
+	int n_mic = cd->config->num_mic_locations;
+	int i;
+
+	sin_az = sin_fixed_16b(Q_SHIFT_LEFT((int32_t)az, 12, 28)); /* Q1.15 */
+	cos_az = cos_fixed_16b(Q_SHIFT_LEFT((int32_t)az, 12, 28)); /* Q1.15 */
+	src_x = Q_MULTSR_32X32((int32_t)cos_az, SOURCE_DISTANCE, 15, 12, 12);
+	src_y = Q_MULTSR_32X32((int32_t)sin_az, SOURCE_DISTANCE, 15, 12, 12);
+
+	for (i = 0; i < n_mic; i++)
+		d[i] = distance_from_source(cd, i, src_x, src_y, 0);
+
+	for (i = 0; i < n_mic - 1; i++) {
+		delta_d = d[i + 1] - d[0]; /* Meters Q4.12 */
+		cd->direction.timediff_iter[i] =
+			(int32_t)((((int64_t)delta_d) << 19) / SPEED_OF_SOUND);
+	}
+}
+
+static inline int64_t mean_square_time_difference_err(struct tdfb_comp_data *cd)
+{
+	int64_t err = 0;
+	int32_t delta;
+	int i;
+
+	for (i = 0; i < cd->config->num_mic_locations - 1; i++) {
+		delta = cd->direction.timediff[i] - cd->direction.timediff_iter[i];
+		err +=  (int64_t)delta * delta;
+	}
+
+	return err;
+}
+
+static int unwrap_radians(int radians)
+{
+	int a = radians;
+
+	if (a > PI_Q12)
+		a -= PIMUL2_Q12;
+
+	if (a < -PI_Q12)
+		a += PIMUL2_Q12;
+
+	return a;
+}
+
+static inline void iterate_source_angle(struct tdfb_comp_data *cd)
+{
+	int64_t err_prev;
+	int64_t err;
+	int32_t ds1;
+	int32_t ds2;
+	int az_step;
+	int az_slow;
+	int az;
+	int i;
+
+	/* Get theoretical time differences for previous angle */
+	az = cd->direction.az_slow;
+	theoretical_time_differences(cd, az);
+	err_prev = mean_square_time_difference_err(cd);
```

Comment:
ditto here

Reply:
Move err_prev before theoretical_time_differences() is not  OK because the cd->direction.timediff_iter[] would be based on previous copy() round.